core/consensus/babe/src/lib.rs

```diff
@@ -75,28 +155,918 @@ impl Decode for BabeSeal {
+		debug!(target: "babe", "checkpoint 2");
+
+		net.start();
+		debug!(target: "babe", "checkpoint 3");
+
+		let peers = &[
+			(0, Keyring::Alice),
+			(1, Keyring::Bob),
+			(2, Keyring::Charlie),
+		];
+
+		let net = Arc::new(Mutex::new(net));
+		let mut import_notifications = Vec::new();
+		debug!(target: "babe", "checkpoint 4");
+		let mut runtime = current_thread::Runtime::new().unwrap();
+		for (peer_id, key) in peers {
+			let client = net.lock().peer(*peer_id).client().clone();
+			let environ = Arc::new(DummyFactory(client.clone()));
+			import_notifications.push(
+				client.import_notification_stream()
+					.take_while(|n| Ok(!(n.origin != BlockOrigin::Own && n.header.number() < &5)))
+					.for_each(move |_| Ok(()))
+			);
+
+			let config = Config::get_or_compute(&*client)
+				.expect("slot duration available");
+
+			let inherent_data_providers = InherentDataProviders::new();
+			register_babe_inherent_data_provider(
+				&inherent_data_providers, config.get()
+			).expect("Registers babe inherent data provider");
+
+			let babe = start_babe(BabeParams {
+				config,
+				local_key: Arc::new(key.clone().into()),
+				block_import: client.clone(),
+				client,
+				env: environ.clone(),
+				sync_oracle: DummyOracle,
+				on_exit: futures::empty(),
+				inherent_data_providers,
+				force_authoring: false,
+			}).expect("Starts babe");
+
+			runtime.spawn(babe);
+		}
+		debug!(target: "babe", "checkpoint 5");
+
+		// wait for all finalized on each.
+		let wait_for = ::futures::future::join_all(import_notifications)
+			.map(drop)
+			.map_err(drop);
+
+		let drive_to_completion = ::tokio::timer::Interval::new_interval(TEST_ROUTING_INTERVAL)
+			.for_each(move |_| {
+				net.lock().send_import_notifications();
+				net.lock().sync_without_disconnects();
+				Ok(())
+			})
+			.map(drop)
+			.map_err(drop);
+
+		runtime.block_on(wait_for.select(drive_to_completion).map_err(drop)).unwrap();
+	}
+
+	#[test]
+	#[allow(deprecated)]
+	#[should_panic]
+	fn old_seals_rejected() {
+		drop(env_logger::try_init());
+		generic::DigestItem::<network::test::Hash, Public, primitives::sr25519::Signature>::Seal(0, Signature([0; 64])).as_babe_seal().unwrap();
+	}
+
+	#[test]
+	fn wrong_number_rejected() {
+		drop(env_logger::try_init());
+		let bad_seal = generic::DigestItem::<network::test::Hash, Public, primitives::sr25519::Signature>::Consensus([0; 4], Signature([0; 64]).encode());
+		assert!(bad_seal.as_babe_seal().is_none())
+	}
+
+	#[test]
+	#[should_panic]
+	fn bad_seal_rejected() {
+		drop(env_logger::try_init());
+		let bad_seal = generic::DigestItem::<network::test::Hash, Public, primitives::sr25519::Signature>::Consensus(BABE_ENGINE_ID, Signature([0; 64]).encode());
+		bad_seal.as_babe_seal().expect("we should not decode this successfully");
+	}
+
+	#[test]
+	fn can_author_block() {
+		drop(env_logger::try_init());
+		let randomness = &[];
+		let pair = sr25519::Pair::generate();
+		let mut i = 0;
+		loop {
```

Comment:
This test relies on some random inputs (keys) so I am wondering if there is a possibility that it either fails or loops forever on some rare occasions and it may hurt CI

Reply:
Nope.  Secret keys are random 252 bit numbers that all behave essentially the same.  As an aside, there are a few insecure secret key choices, but we safely assume you'll never hit them because 2^252 is astronomically large.